### PR TITLE
repo-tools: detect and delete unnecessary api reports

### PR DESCRIPTION
**Pull request by freben**
The `api-reports` command now checks for api report files that no longer apply. If it finds such files, it's treated basically the same as report errors do, and the check fails.

For example, if you had an `api-report-alpha.md` but then removed the alpha export, the reports generator would now report that this file needs to be deleted.

### CODE REVIEW

File: packages/repo-tools/src/commands/api-reports/api-extractor.ts [hunk 1]
```diff
@@ -363,7 +364,9 @@ export async function runApiExtraction({
   }
   const warnings = new Array<string>();
 
-  for (const { packageDir, name } of packageEntryPoints) {
+  for (const [packageDir, group] of Object.entries(
+    groupBy(packageEntryPoints, ep => ep.packageDir),
+  )) {
```
Comment (freben) [inline packages/repo-tools/src/commands/api-reports/api-extractor.ts:369]: this change of the main for loop could in theory be avoided, but it fits in with the new functionality and has the benfit that the command now no longer prints out the "Processing ${packageDir}" message multiple times when there are multiple exports.

File: packages/repo-tools/src/commands/api-reports/api-extractor.ts [hunk 2]
```diff
@@ -374,209 +377,237 @@ export async function runApiExtraction({
...
+        );
+      }
+      if (warningCountAfter > warningCountBefore) {
+        warnings.push(
+          `The API Report for ${packageDir} introduces new warnings. ` +
+            'Please fix these warnings in order to keep the API Reports tidy.',
+        );
+      }
+      if (staleReportFiles.length) {
+        if (noBail) {
```
Comment (Rugvip) [inline packages/repo-tools/src/commands/api-reports/api-extractor.ts:597]:
```suggestion
        if (isLocalBuild) {
```

Thinking this is better? Just deletes the file when running locally, but errors out in CI.

Thinking that this should be after the loop as well?

File: packages/repo-tools/src/commands/api-reports/api-extractor.ts [hunk 3]
```diff
@@ -374,209 +377,237 @@ export async function runApiExtraction({
...
-    const suffix = name === 'index' ? '' : `-${name}`;
-    const reportFileName = `api-report${suffix}.md`;
-    const reportPath = resolvePath(projectFolder, reportFileName);
-
-    const warningCountBefore = await countApiReportWarnings(reportPath);
+    const staleReportFiles = fs.readdirSync(projectFolder).filter(filename => {
+      const match =
+        filename.match(/^(.+)-api-report\.md$/) ||
+        filename.match(/^api-report-(.+)\.md$/);
+      return match && !names.includes(match[1]);
```
Comment (Rugvip) [inline packages/repo-tools/src/commands/api-reports/api-extractor.ts:386]: This doesn't check whether `api-report.md` should exist. I'm thinking it's better to filter these based on the actual `reportFileName` below? Could make it a set here, then `staleReportFiles.delete(reportFileName)` in the loop, and check that it's empty afterwards.

**Review by Rugvip — CHANGES_REQUESTED** [review]
Nice! 👍, just a some nits

File: packages/repo-tools/src/commands/api-reports/api-extractor.ts (hunk 2 above)
Comment (freben) [reply packages/repo-tools/src/commands/api-reports/api-extractor.ts]: quite right! updated and moved

File: packages/repo-tools/src/commands/api-reports/api-extractor.ts (hunk 3 above)
Comment (freben) [reply packages/repo-tools/src/commands/api-reports/api-extractor.ts]: Aight, that does make things a bit neater too, fixed

**Review by Rugvip — APPROVED** [review]
:shipit:

**Pull request merged**